### ISSUE TITLE
https://github.com/scikit-learn/scikit-learn/issues/14735

### ISSUE THREAD

**Issue by Sandy4321** [jrzaurin/LightGBM-with-Focal-Loss#2]
great post and code
but what is about
scale_pos_weight is used to adjust the imbalance will spoil probability?

https://github.com/microsoft/LightGBM/issues/2107

https://github.com/scikit-learn/scikit-learn/issues/14735

by the way 
did you found really unbalanced data?

**Comment by Sandy4321** [comment]
as it is written : result in poor estimates of the individual class probabilities
https://lightgbm.readthedocs.io/en/latest/Parameters.html

is_unbalance 🔗︎, default = false, type = bool, aliases: unbalance, unbalanced_sets

used only in binary and multiclassova applications

set this to true if training data are unbalanced

Note: while enabling this should increase the overall performance metric of your model, it will also result in poor estimates of the individual class probabilities

**Comment by Sandy4321** [comment]
or
consider performing probability calibration (https://scikit-learn.org/stable/modules/calibration.html)

https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html
class_weight (dict, 'balanced' or None, optional (default=None)) – Weights associated with classes in the form {class_label: weight}. Use this parameter only for multi-class classification task; for binary classification task you may use is_unbalance or scale_pos_weight parameters. Note, that the usage of all these parameters will result in poor estimates of the individual class probabilities. You may want to consider performing probability calibration (https://scikit-learn.org/stable/modules/calibration.html) of your model. The ‘balanced’ mode uses the values of y to automatically adjust weights inversely proportional to class frequencies in the input data as n_samples / (n_classes * np.bincount(y)). If None, all classes are supposed to have weight one. Note, that these weights will be multiplied with sample_weight (passed through the fit method) if sample_weight is specified

**Comment by jrzaurin** [comment]
Hey @Sandy4321 sorry I am a bit late to the issue. Yeah, I am aware of most of what you write up there. Regarding to the individual probabilities, in the post I am comparing performance metrics, and there `in_unbalance` should help. I also wanted to illustrate how it might not be the best to use both, `is_unbalance` and the focal loss. 

Regarding to prob calibration, I must admit that, in my experience, with one exception, never lead to a notable increase in overall metrics (at least when using isotonic regression). It was always better to take the probabilities coming from the algo and compute the prob threshold that lead to the higher F-score

And finally, no, unfortunately I did not find true challenging imbalanced datasets. However, I had two colleagues trying it in two banks for fraud detection and money laundering and the performance increase was quite notable. And in my case, I use it for risk prediction in insurance an the increase in F1 score was also high (~5%).

**Comment by Sandy4321** [comment]
great thanks for answer
1
what you think
Are you using the parameter is_unbalanced from Light GBM parameters ?
I was trying to compare the results when I put the parameter to False or True but I get quite the same result (LB and CV)…
https://www.kaggle.com/c/home-credit-default-risk/discussion/62515

As you know the data are quite unbalanced so I don't know if we should use it !
maybe you can try this data

Please do hyperparameter tuning for scaleposweight between (1 - 11.38) and it will give you a significant increase in CV score

or

I have no idea whether this affects your result, but it has been debated in previous imbalanced data competitions that setting clf.set_params(is_unbalance=True) is probably not a good idea. This setting will take the actual class imbalance ratio (11.39) as a multiplier for 1 class. It is debatable whether the ~10x imbalance needs any correction, but it almost certainly is not optimal to use the actual imbalance ratio.

If you really want to explore this, an alternative option is to set is_unbalance=False and treat scale_pos_weight as a hyperparameter to optimize in the range 1 - 11.39
https://www.kaggle.com/mlisovyi/modular-good-fun-with-ligthgbm/comments#337494

2
https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html
it is messy description : what exactly to set to treat un balanced data
class_weight or is_unbalance = True or use scale_pos_weight
class_weight (dict, 'balanced' or None, optional (default=None)) – Weights associated with classes in the form {class_label: weight}. Use this parameter only for multi-class classification task; for binary classification task you may use is_unbalance or scale_pos_weight parameters

may explain what you used exactly ?
better to share code

**Comment by Sandy4321** [comment]
though if lighbm crashes conceptually  (classes are not well-separable )  then nothing can help
https://www.kaggle.com/mlisovyi/modular-good-fun-with-ligthgbm/comments#337494
As I said in my original comment, it is debatable whether ~10 imbalance needs any correction. Looks like you found out the same on your own. I included scale_pos_weight in one of my grid searches and it came out at 2.24 and 1.07 in top two solutions for which CV scores differ by 0.00012. Safe to say that is_unbalance=True should not be used. Same for scale_pos_weight, or maybe use a value just slightly larger than 1.
Also, a more general problem seems to be that data classes are not well-separable to begin with - see here. In such cases weighting or sampling data most likely will not work.

**Comment by Sandy4321** [comment]
by the you can generate data by yourself
with any unbalance ratio
https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_classification.html
weights   : array-like of shape (n_classes,) or (n_classes - 1,), (default=None)
The proportions of samples assigned to each class. If None, then classes are balanced. Note that if len(weights) == n_classes - 1, then the last class weight is automatically inferred. More than n_samples samples may be returned if the sum of weights exceeds 1

**Comment by Sandy4321** [comment]
great new I found data for you
use this link
https://machinelearningmastery.com/imbalanced-classification-with-the-fraudulent-credit-card-transactions-dataset/

and
https://raw.githubusercontent.com/jbrownlee/Datasets/master/creditcard.csv.zip

also code
http://rstudio-pubs-static.s3.amazonaws.com/334864_28050f7860dd4927a596872f0cd52401.html

and
https://github.com/pawellos1234/credit-card-fraud

lets see if your code can do better?

**Comment by Sandy4321** [comment]
also you can get data here
https://www.kaggle.com/mlg-ulb/creditcardfraud

**Comment by Sandy4321** [comment]
and a lot of code to compare with
https://www.kaggle.com/mlg-ulb/creditcardfraud/kernels

**Comment by jrzaurin** [comment]
Hi @Sandy4321 

I am a bit loss as to what do you want here (?)

Let me go step by step: 

1) in the post I discussed the parameter `is_unbalance ` to illustrate what it does, mathematically (See eq 4 in the post) and how use it with the focal loss might not be a good idea 

2) `scale_pos_weight` is a parameter that should not be tuned. `scale_pos_weight` scales the minority class during the computation of the loss. This scaling depends on the real ratio between minor/major classes and it should not be tuned. It is what it is. The question one should answer (depending on whether you want precision, recall, f-score) is if you want to add that weight or not, but `scale_pos_weight` should, a priori, not be tuned. 

3) you mention: _"may explain what you used exactly ? better to share code"_ are you asking me to share some code? you are writing in the repo...all the code is here. Also you recommend me the link to the lightGBM documentation. As you can imagine, I have been there many times...I use lightGBM "all the time". Maybe I am missing something...

4) On another note, note that due to the nature of what `scale_pos_weight` does, there is no guarantee that the algorithm performs better on the test set. As mentioned before, it depends on what is it that one is prioritising at test time, i.e. your success metric.

5) thanks for the resources that you send me, I am of course familiar with Jason's work (I think anyone in the space is) and I of course know how to generate data with Sklearn. Again...are you asking me something or simply adding resources? Also the credit fraud detection data is what I use in the post. Is synthetically generated and a couple of columns are highly correlated with the target, making it a really easy data, not really challenging for lightGBM with/without the focal loss. 

If you do not mind I am going to go ahead and close the issue. If there is a further issue or something you want to ask and is required, I will reopen. 

Thanks!

**Comment by Sandy4321** [comment]
Dear Javier
Thank you very much for taking care
sorry for misunderstanding
1
I meant exactly this mess with using scale_pos_weight vs is_unbalance vs class_weight
and asked you what you used as baseline to compare with Focal-Loss
For example in these link to kaggle competition they surprised what and how to use
2
you meant that there is no data (no real data datasets with high unbalance) - so I asked you : why not to simulate data ?
simulated data may be even better than   real data - we know exactly what to find
the same idea like
https://maciejkula.github.io/spotlight/datasets/synthetic.html
generating synthetic datasets with known properties, for model testing and experimentation.
3
 
sorry for confusion with creditcardfraud data...
another idea how to find dataset - may be to use RTB data
in RTB case there are should be much less clicks then no clicks
again
Thank you very much for answers

**Comment by jrzaurin** [comment]
_why not to simulate data ?_:  when you simulate data, like the one at the Kaggle Credict Card Fraud competition, you set which features are actually predictive. In that scenario, even if the data is highly imbalanced, it is really easy to predict because any algorithm would find those synthetic correlations. 

The link you post is for the great work by Maciej Kula, and his `spotlight` package. That method `generate_sequential` is designed by sequential data, not something I really wanted to do in this repo and the corresponding post.

Cheers
J.
